pallets/stateful-storage/src/signed_extension.rs

```diff
@@ -0,0 +1,200 @@
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, SignedExtension},
+	transaction_validity::{InvalidTransaction, TransactionValidity, TransactionValidityError},
+	DispatchError,
+};
+
+/// The SignedExtension trait is implemented on CheckFreeExtrinsicUse to validate the request. The
+/// purpose of this is to ensure that the target_hash is verified in transaction pool before getting
+/// into block. This is to reduce the chance that capacity consumption due to stale hash
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct StatefulStorageSignedExtension<T: Config + Send + Sync>(PhantomData<T>);
+
+impl<T: Config + Send + Sync> StatefulStorageSignedExtension<T> {
+	/// Create new `SignedExtension`.
+	pub fn new() -> Self {
+		Self(sp_std::marker::PhantomData)
+	}
+
+	/// Verifies the hashes for an Itemized Stateful Storage extrinsic
+	fn verify_hash_itemized(
+		msa_id: &MessageSourceId,
+		schema_id: &SchemaId,
+		target_hash: &PageHash,
+	) -> TransactionValidity {
+		const TAG_PREFIX: &str = "StatefulStorageHashItemized";
+
+		if let Ok(Some(page)) = Pallet::<T>::get_itemized_page_for(*msa_id, *schema_id) {
+			let current_hash: PageHash = page.get_hash();
+
+			ensure!(
+				&current_hash == target_hash,
+				map_dispatch_error(Error::<T>::StalePageState.into())
+			);
+
+			return ValidTransaction::with_tag_prefix(TAG_PREFIX)
+				.and_provides((msa_id, schema_id))
+				.build()
+		}
+		Ok(Default::default())
+	}
+
+	/// Verifies the hashes for a Paginated Stateful Storage extrinsic
+	fn verify_hash_paginated(
+		msa_id: &MessageSourceId,
+		schema_id: &SchemaId,
+		page_id: &PageId,
+		target_hash: &PageHash,
+	) -> TransactionValidity {
+		const TAG_PREFIX: &str = "StatefulStorageHashPaginated";
+		if let Ok(Some(page)) = Pallet::<T>::get_paginated_page_for(*msa_id, *schema_id, *page_id) {
+			let current_hash: PageHash = page.get_hash();
+
+			ensure!(
+				&current_hash == target_hash,
+				map_dispatch_error(Error::<T>::StalePageState.into())
+			);
+
+			return ValidTransaction::with_tag_prefix(TAG_PREFIX)
+				.and_provides((msa_id, schema_id, page_id))
+				.build()
+		}
+
+		Ok(Default::default())
+	}
+}
+
+impl<T: Config + Send + Sync> sp_std::fmt::Debug for StatefulStorageSignedExtension<T> {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "StatefulStorageSignedExtension<{:?}>", self.0)
+	}
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+/// Map a module DispatchError to an InvalidTransaction::Custom error
+pub fn map_dispatch_error(err: DispatchError) -> InvalidTransaction {
+	InvalidTransaction::Custom(match err {
+		DispatchError::Module(module_err) =>
+			<u32 as Decode>::decode(&mut module_err.error.as_slice())
+				.unwrap_or_default()
+				.try_into()
+				.unwrap_or_default(),
+		_ => 255u8,
```

Comment:
It looks like we might end up with an error code that isn’t entirely accurate. Same in matching case - zero.

Reply:
Also, just a note: the "Custom Error" code that we get as part of the RPC "1010" error is entirely arbitrary, and does not indicate the pallet. So a code of `9` (StalePageHash) is not meaningful to the end user, because they don't know what pallet enum that comes from. I do think somewhere there is an enum of custom RPC error codes for Frequency; perhaps we should map to that instead? Sadly Polkadot/Substrate does not give us a way to propagate anything more than an integer value for custom RPC errors.

Reply:
Interesting approach but works.

Reply:
Probably a wider issue than this PR, but the universal enum idea might not be so bad; we wouldn't need to map every error from every pallet; only the ones that can happen in a SignedExtension check. There shouldn't be that many.

Anyway, this approach works for now; maybe it's worth a backlog issue to revisit how we do this across all of our signed extensions, or maybe not.